### ISSUE TITLE
Localization issues

### ISSUE THREAD

**Issue by Hatsushira** [carekit-apple/CareKit#602]
Dear all,

I had a break on working with CareKit for a while, so it's possible this issue exists for longer than just now. I'm currently on Xcode 12.5 and on the main branch. I linked the CareKit libraries with the github link to source code and the libraries are linked and working. What happens to me now is that the UI elements of CareKit aren't translated anymore. But this works in an older version of CareKit (approx Dec. 2020) on my parent's iPad.

I switched on the missing phrases option in Xcode and this is on of the error messages I'm getting.

`2021-06-14 20:39:04.583255+0200 PHR[7363:110426] [strings] ERROR: Log not found in table Localizable of bundle CFBundle 0x7fb887463bc0 </Users/gerdgrossmann/Library/Developer/CoreSimulator/Devices/BD1C21E8-9CC2-4E46-93C0-628CDB6435D0/data/Containers/Bundle/Application/4D7CEFDB-13D8-4992-BD59-768E600BCB3A/PHR.app/CareKit_CareKitUI.bundle> (not loaded)`

`Log` not found means that it is actually processed by the function, because the phrase is "LOG" but it's not translated. So could you help me with any ideas or is it perhaps related to the following commit #566 ?

**Comment by Hatsushira** [comment]
I use SPM with your first option described in your readme. I add a dependency to the main branch.
Currently I watch this issue on the OCKButtonLogTaskView. The phrase LOG is "translated to Log, COMPLETED to Completed and the same goes for the options if you double click on a completed entry to remove it.
Additiontally, I have custom versions of the OCKButtonLogTaskView, all my added fields labels etc. are translated, but the Log Button is not.

**Comment by Hatsushira** [comment]
Yes, in German.
Yes, I do have a `Localizable.strings` file and a Localizable.stringsdict. I copied all the phrases in these files as suggested.
My app is translated in all other areas. I only experience this problem with the button log. I attached some images for you. As you can see in the second picture the phrases of the stringsdict are used. On the button log they're not.
[](https://ibb.co/gwD3QTH)
[](https://ibb.co/0BBNcnV)

**Comment by erik-apple** [comment]
The fact that most of your translations are being picked up leads me to believe that your localizations bundles are being discovered as intended. `"LOG"` and `"COMPLETED"` appear to be the correct keys as of the latest commit. Can you confirm that your German localization files contain they correct keys? For example, you might have "COMPLETE" instead of "COMPLETED".

**Comment by Hatsushira** [comment]
Unfortunately, yes. They are all in. These are my contents for the CareKit Part.
```
// MARK: CareKitUI
"ADDRESS" = "Adresse";
"ANNOUNCE_EVENT_DELETED" = "Aufgezeichnetes Event gelöscht";
"BEGIN" = "Start";
"CANCEL" = "Abbrechen";
"CALL" = "Anrufen";
"CLOSE" = "Schließen";
"COMPLETED" = "Abgeschlossen";
"CONTACTS" = "Kontakte";
"DELETE" = "Löschen";
"DONE" = "Fertig";
"DOUBLE_TAP_MAP" = "Doppelt Tippen um auf der Karte anzuzeigen";
"DOUBLE_TAP_TO_COMPLETE" = "Doppelt Tippen um Abzuschließen";
"DOUBLE_TAP_TO_INCOMPLETE" = "Doppelt Tippen um es zu aktivieren";
"DOUBLE_TAP_TO_RECORD_EVENT" = "Doppelt Tippen um ein Event zu markieren";
"DOUBLE_TAP_TO_REMOVE_EVENT" = "Doppelt Tippen um das Event zu löschen";
"EMAIL" = "E-Mail";
"EVENT" = "Event";
"GOAL" = "Ziel";
"HIGH" = "Hoch";
"INCOMPLETE" = "Nicht abgeschlossen";
"LOG" = "Aufzeichen";
"LOG_ENTRY" = "Eintrag";
"LOGGING" = "Aufzeichnen";
"LOW" = "Niedrig";
"MARK_COMPLETE" = "Als abgeschlossen markieren";
"MESSAGE" = "Nachricht";
"NO_DATA" = "Keine Daten";
"NO_TASKS" = "Keine Aufgaben";
"NO_EVENTS" = "Keine Events";
"PROGRESS" = "Fortschritt";
"SEARCH" = "Suchen";
"SELECTED" = "Ausgewählt";
"TASKS" = "Aufgaben";
"THREE_FINGER_SWIPE_DAY" = "Mit drei Fingern wischen um zu nächsten oder vorherigen Tag zu gelangen";
"THREE_FINGER_SWIPE_WEEK" = "Mit drei Fingern wischen um zur nächsten oder vorherige Woche zu gelangen";
"TODAY" = "Heute";
```

**Comment by erik-apple** [comment]
Thanks for providing your translations! That should help a lot.
We'll start digging into this a bit more.

**Comment by Hatsushira** [comment]
Thank you Erik, I really appreciate your help. Just because I'm curious I switched my project back from `main` to `master` branch and what can I say. There it works. Switching back to `main` the phrases aren't translated anymore.
https://ibb.co/CtjvGNf

**Comment by gavirawson-apple** [comment]
Hi @Hatsushira! I tried to reproduce the issue you're seeing, but haven't had any luck. On the `main` branch, the strings seem to be displaying in German as expected:

<img src="https://user-images.githubusercontent.com/51756298/123494741-bf598f00-d5d5-11eb-9178-c16f0b4a7cdc.png" width="300px">

I wonder if there's something I'm missing here that's causing the underlying issue. Are you still seeing an issue on your end? If so, what was the process you went through to add the German strings to the project?

**Comment by Hatsushira** [comment]
Hi @gavirawson-apple thank you for you help Yes, I'm still facing this issue.
I added a group Localization to my main project and added `Localizable.strings` and `Localizable.stringsdict` to it. Then for Localization I chose English and German. Then I copied the phrases from CareKit into those files and translated them. Basically, that's it.

**Comment by gavirawson-apple** [comment]
Gotcha, thanks for the info. Was the process you went through somewhat like like the instructions [here](https://developer.apple.com/documentation/xcode/adding-support-for-languages-and-regions)?

**Comment by Hatsushira** [comment]
Yes, I'm sorry, that I'm not very helpful. The weird thing is that when I switch back to the master branch here from this repository it does work. The only change is the repository, my code stays the same. Perhaps it has something to do with my directory structure?
https://imgur.com/mqwexQv

**Comment by gavirawson-apple** [comment]
No worries at all! All the information you've been giving has been super helpful. I'm trying to think how our setups might be different, I'm also testing the localization on Xcode 12.5.

I think you're on to something with the directory structure. Mine looks a bit different from yours:

<img width="400" alt="Screen Shot 2021-07-01 at 12 50 48 PM" src="https://user-images.githubusercontent.com/51756298/124182034-2e6a3400-da6b-11eb-96e4-8e72403856d1.png">

I'd recommend starting from scratch and seeing if the issue is still there, is that possible on your end? You can add a German strings file by navigating to and tapping the circled "plus" button below:

<img width="1521" alt="Screen Shot 2021-07-01 at 12 52 35 PM" src="https://user-images.githubusercontent.com/51756298/124182239-71c4a280-da6b-11eb-918a-63bcf58c39e4.png">

**Comment by Hatsushira** [comment]
So, I created a complete new Xcode project, with only the standard Base and English as development language, added my files and was able to run the project. I enabled the missing localizable strings in the run options. Then I run the projects and I got all the missing phrases warnings as expected.
Then I added a localizable.strings only for English. All phrases were translated but for the Log phrase I got the warning that it's missing  in the package. I think for the Button Log CareKit only looks in the CareKit package for the phrase and not in my package. CareKit is only linked and it's read only for me. I just added the repository as link. 

This is my project overview. 

https://imgur.com/i2nf7Ax

The only change after switching from master to main branch with a reference to language is this commit. 
https://github.com/carekit-apple/CareKit/commit/7dfcdef4de7e91449e86ac963043b44188a6ba31

**Comment by gavirawson-apple** [comment]
Thanks for all of the info! I believe the source of the issue has something to do with SPM. We're investigating and will let you know when we have a fix. In the meantime, if you'd like, you can embed CareKit in your app rather than using SPM. That will allow you to add a German strings file directly to the framework, which works on my end.

**Comment by Hatsushira** [comment]
Thank you for your assistance.

**Comment by Hatsushira** [comment]
I have additional information:
I cloned this repo on my own server. Added localization files to the localization folder of CareKitUI and I get the expected behaviour, that the phrase actually **is translated to German in the user interface**. BUT: I still get the error in the console:
`2021-07-18 22:16:14.589849+0200 PHR[59554:606556] [strings] ERROR: Log not found in table Localizable of bundle CFBundle 0x7ff086d3dd80 </Users/gerdgrossmann/Library/Developer/CoreSimulator/Devices/14DB5404-A621-4377-9283-D01079AFA810/data/Containers/Bundle/Application/869B73F3-A3B3-42E2-A2F8-119BB3C51804/PHR.app/CareKit_CareKitUI.bundle> (not loaded)`

**Comment by gavirawson-apple** [comment]
That is very strange, I haven't seen that one before! Perhaps it only happens in the simulator.

**Comment by Hatsushira** [comment]
Me, neither. Perhaps my project is somehow messed up.

I got my localization now working and can move on. Thanks again for your patience.

**Comment by gavirawson-apple** [comment]
Of course! Glad things are working. It may take some time, but we will solve this SPM+Loc issue on our end!

**Comment by zxcheergo** [comment]
How do you get the Localizable.strings file ?

**Comment by zxcheergo** [comment]
Is there any tutorial on how to add the localization on CareKit? Thanks!
